### PR TITLE
ARROW-2577: [Plasma] Add asv benchmarks for plasma

### PR DESCRIPTION
**Pull request by pcmoritz**
This adds some initial ASV benchmarks for plasma:

- Put latency
- Get latency
- Put throughput for 1KB, 10KB, 100KB, 1MB, 10MB, 100MB

It also includes some minor code restructuring to expose the start_plasma_store method.

### CODE REVIEW

File: python/benchmarks/plasma.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class PlasmaBenchmarks(object):
+
+    def setup(self):
+        self.plasma_store_ctx = plasma.start_plasma_store(plasma_store_memory=10**9)
+        plasma_store_name, p = self.plasma_store_ctx.__enter__()
+        self.plasma_client = plasma.connect(plasma_store_name, "", 64)
+
+        self.data_1kb = np.random.randn(1000 // 8)
```
Comment (pitrou) [inline python/benchmarks/plasma.py]: You should use parametrization for the various sizes (you can look at the conversion benchmarks to see how that is done, or see [docs](https://asv.readthedocs.io/en/latest/writing_benchmarks.html#parameterized-benchmarks)). Also I don't think we should test so many sizes.  Testing a very small size (1kb) and a large-ish size (10mb) sounds sufficient.

File: python/benchmarks/plasma.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import numpy as np
+
+import pyarrow as pa
+import pyarrow.plasma as plasma
+
+from . import common
+
+
+class PlasmaBenchmarks(object):
```
Comment (pitrou) [inline python/benchmarks/plasma.py]: From a high-level point of view, what do we want to benchmark here? The plasma client, or client + store? You may want to choose an explicit timer that reflects that decision (see the `timer` attribute [here](https://asv.readthedocs.io/en/latest/writing_benchmarks.html#timing)). Also add a docstring :-)

**Review by pitrou — COMMENTED** [review]
Looks nice :-) Some comments below.

File: python/benchmarks/plasma.py (hunk 2 above)
Comment (pcmoritz) [reply python/benchmarks/plasma.py]: I added a docstring, not sure what you mean with the timer attribute; the default seems fine to me :)

Comment (pitrou) [reply python/benchmarks/plasma.py]: I mean that if you want to time the client overhead, you need a timer that measures the elapsed CPU time of the current process is required (but that is `asv`'s default, AFAICT). If OTOH you want to measure the whole client + server roundtrip, you need a timer that measures the elapsed wall clock time.

Comment (pcmoritz) [reply python/benchmarks/plasma.py]: Oh I see, thanks for pointing that out. I do want to measure the roundtrip wallclock time, and think timeit which is the default for asv is doing that (according to https://docs.python.org/3.0/library/timeit.html).

Comment (pitrou) [reply python/benchmarks/plasma.py]: Actually, according to https://asv.readthedocs.io/en/latest/writing_benchmarks.html#timing the default is process CPU time (by way of `time.process_time`). Which makes sense in most cases but not here :-)

Comment (pcmoritz) [reply python/benchmarks/plasma.py]: Thanks, I didn't realize they were overriding the timeit default, should be fixed now :)

**Comment by pcmoritz** [comment]
@pitrou Is there a way to get all the numbers for each parameter from a command line run? For me it only shows the first one and the rest is omitted with ... (at the very right):

```
ubuntu@ip-172-31-49-70:~/arrow/python$ asv run --python=same
· Discovering benchmarks
· Running 14 total benchmarks (1 commits * 1 environments * 14 benchmarks)
[  0.00%] ·· Building for existing-py_home_ubuntu_anaconda3_bin_python
[  0.00%] ·· Benchmarking existing-py_home_ubuntu_anaconda3_bin_python
[  7.14%] ··· Running array_ops.ScalarAccess.time_as_py                                                                                                                        9.41ms
[ 14.29%] ··· Running array_ops.ScalarAccess.time_getitem                                                                                                                     41.09ms
[ 21.43%] ··· Running convert_builtins.ConvertArrayToPyList.time_convert                                                                                                  41.60ms;...
[ 28.57%] ··· Running convert_builtins.ConvertPyListToArray.time_convert                                                                                                  15.41ms;...
[ 35.71%] ··· Running convert_builtins.InferPyListToArray.time_infer                                                                                                      19.75ms;...
[ 42.86%] ··· Running convert_pandas.PandasConversionsFromArrow.time_to_series                                                                                             1.28ms;...
[ 50.00%] ··· Running convert_pandas.PandasConversionsToArrow.time_from_series                                                                                           594.60μs;...
[ 57.14%] ··· Running convert_pandas.ZeroCopyPandasRead.time_deserialize_from_buffer                                                                                         609.45μs
[ 64.29%] ··· Running convert_pandas.ZeroCopyPandasRead.time_deserialize_from_components                                                                                     597.48μs
[ 71.43%] ··· Running microbenchmarks.PandasObjectIsNull.time_PandasObjectIsNull                                                                                           2.76ms;...
[ 78.57%] ··· Running plasma.SimplePlasmaLatency.time_plasma_put                                                                                                             500.94ms
[ 85.71%] ··· Running plasma.SimplePlasmaLatency.time_plasma_putget                                                                                                          691.88ms
[ 92.86%] ··· Running plasma.SimplePlasmaThroughput.time_plasma_put_data                                                                                                 562.55μs;...
[100.00%] ··· Running streaming.StreamReader.time_read_to_dataframe                                                                                                      232.14ms;...
```

**Comment by pitrou** [comment]
@pcmoritz The `-e` flag should do it, IIRC.

**Pull request closed without merging**